src/main/java/org/apache/sysml/runtime/instructions/gpu/context/GPUMemoryManager.java

```diff
@@ -0,0 +1,561 @@
+		if(LOG.isDebugEnabled())
+			LOG.debug("Removing the GPU object: " + gpuObj);
+		allocatedGPUObjects.removeIf(a -> a.equals(gpuObj));
+	}
+	
+	
+	/**
+	 * Get size of allocated GPU Pointer
+	 * @param ptr pointer to get size of
+	 * @return either the size or -1 if no such pointer exists
+	 */
+	public long getSizeAllocatedGPUPointer(Pointer ptr) {
+		if(allocatedGPUPointers.containsKey(ptr)) {
+			return allocatedGPUPointers.get(ptr);
+		}
+		return -1;
+	}
+	
+	public GPUMemoryManager(GPUContext gpuCtx) {
+		long free[] = { 0 };
+		long total[] = { 0 };
+		cudaMemGetInfo(free, total);
+		if(free[0] < 0.7*total[0]) {
+			LOG.warn("Potential under-utilization: GPU memory - Total: " + (total[0] * (1e-6)) + " MB, Available: " + (free[0] * (1e-6)) + " MB on " + gpuCtx 
+					+ ". This can happen if there are other processes running on the GPU at the same time.");
+		}
+		else
+			LOG.info("GPU memory - Total: " + (total[0] * (1e-6)) + " MB, Available: " + (free[0] * (1e-6)) + " MB on " + gpuCtx);
+		if (GPUContextPool.initialGPUMemBudget() > OptimizerUtils.getLocalMemBudget()) {
+			LOG.warn("Potential under-utilization: GPU memory (" + GPUContextPool.initialGPUMemBudget()
+					+ ") > driver memory budget (" + OptimizerUtils.getLocalMemBudget() + "). "
+					+ "Consider increasing the driver memory budget.");
+		}
+	}
+	
+	/**
+	 * Invoke cudaMalloc
+	 * 
+	 * @param A pointer
+	 * @param size size in bytes
+	 * @return allocated pointer
+	 */
+	private Pointer cudaMallocWarnIfFails(Pointer A, long size) {
+		try {
+			cudaMalloc(A, size);
+			allocatedGPUPointers.put(A, size);
+			return A;
+		} catch(jcuda.CudaException e) {
+			LOG.warn("cudaMalloc failed immediately after cudaMemGetInfo reported that memory of size " + size + " is available. "
+					+ "This usually happens if there are external programs trying to grab on to memory in parallel.");
+			return null;
+		}
+	}
+	
+	/**
+	 * Allocate pointer of the given size in bytes.
+	 * 
+	 * @param opcode instruction name
+	 * @param size size in bytes
+	 * @return allocated pointer
+	 * @throws DMLRuntimeException if error
+	 */
+	public Pointer malloc(String opcode, long size) throws DMLRuntimeException {
+		if(size < 0) {
+			throw new DMLRuntimeException("Cannot allocate memory of size " + size);
+		}
+		long t0 = DMLScript.STATISTICS ? System.nanoTime() : 0;
+		// Step 1: First try reusing exact match in rmvarGPUPointers to avoid holes in the GPU memory
+		Pointer A = getRmvarPointer(opcode, size);
+		
+		// Step 2: Allocate a new pointer in the GPU memory (since memory is available)
+		if(A == null && size <= getAvailableMemory()) {
+			A = cudaMallocWarnIfFails(new Pointer(), size);
+			if(LOG.isTraceEnabled()) {
+				if(A == null)
+					LOG.trace("Couldnot allocate a new pointer in the GPU memory:" + size);
+				else
+					LOG.trace("Allocated a new pointer in the GPU memory:" + size);
+			}
+		}
+		
+		// Step 3: Try reusing non-exact match entry of rmvarGPUPointers
```

Comment:
In steps 3 & 4 if a particular block requests size `SIZE` number of bytes, then either of 2 things happen.
Either `SIZE` is fit into the smallest block that it can fit into or all the rmvar-ed blocks are cleared out and space is made. This is rather wasteful since it is possible that 2 or 3 blocks together could have cleared out enough space for a block of size `SIZE` to fit. 

This was being handled by the previous approach. If the LRU approach to freeing up `rmvar-ed` blocks seemed useless, that could be altered with very few lines of code change.

Reply:
There is one optimization that is missing: minimize the number of cleanup of rmvar-ed blocks. We can introduce that if we indeed have an experimental evidence to support it.

The refactoring here has three simple assumptions: 
1. Step 4 & 5: Evictions of matrix blocks are expensive (as they might lead them to be written to disk in case of smaller CPU budget) than doing cuda free/malloc/memset. So, rmvar-ing every blocks is preferred to eviction 
2. Step 3 & 4:  Reusing one rmvar-ed pointer is preferred to reusing multiple pointers.
3. Step 3: When reusing only one rmvar-ed  pointer, its just better to free up that pointer that is just barely bigger than amount of memory needed. The alternative way: Use the rmvar-ed pointer that assumes LRU, which may imply we may have to free-up multiple or a very large recently used memory pointer.

I am open to revisiting these assumptions if we indeed find a counter-example based on experimental evidence :)

Reply:
Do we have experimental evidence for the 3 simple assumptions you've stated:

1. If I've understood point 1 correctly, evicting matrix blocks is expensive, so lets do it all at once instead of one at a time?

2. Reusing one rmvar-ed pointer is preferred to reusing multiple pointers - true, but the tradeoff is more internal fragmentation.

3. So this is a argument pitting the old way vs the new way. Do we have empirical evidence to support that the previous way didn't work and this way will or vice versa? If not, then both methods need to be around and enabled or disabled via a flag, like we've done for the eviction policy. At least one thing is clear - we suffer more internal fragmentation if we do it the new way and at least intuitively we will get better performance since a single block is being deallocated and then allocated as opposed to multiple.